File: 704-binary-search/704-binary-search.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        // for(int i = 0; i < nums.size(); i++) {
        //     cout << nums[i] << ' ';
        // }
        // cout << endl;
        if(nums.size() == 0) return -1;
        if(nums.size() == 1){
            if (nums[0] == target) return 0;
            return -1;
        }
        int middle = nums.size() / 2;
        // cout << "middle is" << middle << endl;
        if(target == nums[middle]) return middle;
        if (target < nums[middle]) 
        {
            vector<int> new_vec = vector<int>(nums.begin(), nums.begin() + middle);
            int result = search(new_vec, target);
            if(result == -1) return -1;
            return result;
        }
        vector<int> new_vec = vector<int>(nums.begin() + middle, nums.end());
        int result = search(new_vec, target);
        if (result == -1) return -1;
        return  result + middle;
    }
};

File: 704-binary-search/704-binary-search.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        begin = 0
        end = len(nums) - 1
        middle = begin + (end - begin) // 2
        while(begin <= end):
            if nums[middle] == target:
                return middle
            if nums[middle] < target:
                begin = middle + 1
            else:
                end = middle - 1
            middle = begin + (end - begin) // 2
        
        return -1
        
        